torch/library.py

```diff
@@ -0,0 +1,74 @@
+import torch
+from ._ops import OpOverload
+from typing import Set
+import warnings
+
+__all__ = ['extend_library']
+
+# User created fragment libraries to extend existing libraries
+fragments_for_existing_libraries = {}
+
+# Set containing the combination of (namespace, operator, DispatchKey) for which a new kernel has been registered
+# The keys in the set are of the form `namespace + "/" + op_name + "/" + dispatch_key`.
+# This set is maintained to ensure that two libraries don't try to override the exact same functionality to avoid
+# libraries calling into kernels not intended to be called.
+impls: Set[str] = set()
+
+class Library:
+    # kind can be DEF, FRAGMENT (in C++ IMPL)
+    def __init__(self, kind, ns, dispatch_key=""):
+        # also prevent users from creating libraries with existing C++ libraries, e.g., aten
+        # if ns in existing_libraries and kind == "DEF":
+        #     raise ValueError("A library with name '" + ns + "' already exists. "
+        #                      "It's not allowed to have more than one library with the same name")
+        # elif kind == "FRAGMENT":
+        #     self.m = torch._C._dispatch_library(kind, ns, dispatch_key)
+        #     self.ns = ns
+        #     fragments_for_existing_libraries[id(self)] = self
+        #     self.op_impls = set()
+        # else:
+        #     raise ValueError("Unsupported kind: ", kind)
+        self.m = torch._C._dispatch_library(kind, ns, dispatch_key)
+        self.ns = ns
+        fragments_for_existing_libraries[id(self)] = self
+        self.op_impls = set()
+
+    def impl(self, op_name, dispatch_key, fn):
+        if isinstance(op_name, str):
+            name = self.ns + '.' + op_name
+        elif isinstance(op_name, OpOverload):
+            name = op_name.__str__() if op_name.overloadname != 'default' else op_name.overloadpacket.__str__()
+        else:
+            raise RuntimeError("impl should be passed either a name or an OpOverload object as the first argument")
+
+        print("--------", name, self.ns)
+        key = self.ns + "/" + name + "/" + dispatch_key
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            # try:
+            self.m.impl(name, dispatch_key, fn)
+        #     except RuntimeWarning:
+        #         if key in impls:
+        # if key in impls:
+            # raise RuntimeError("This is not allowed since there's already a kernel overriding "
+            #                    + str(getattr(getattr(torch.ops, self.ns), name)) + "'s behavior for " + dispatch_key + " dispatch key.")
+        # else:
+        impls.add(key)
+        self.op_impls.add(key)
+
+
+    def remove(self):
+        for key in self.op_impls:
+            impls.remove(key)
+        del fragments_for_existing_libraries[id(self)]
+        del self.m
```

Comment:
I still think `__del__` should suffice for most of this, I'd rather not have "uninitialized" Library objects floating around (where it's now illegal to do operations on them)

Reply:
I think there is one case where we might not want this. If we want to allow to do:
```
returned = torch.ops.aten.sum.default.impl(my_new_sum, "CPU")
```
Basically allowing you to just override the function you want without having to think about library/impl/extend/ etc

So the question is what is returned here? We can say it is a library object and then yes we can bind the lifetime of the function to the library object.
But if we want to allow people to just override it and forget, `returned` can be a handle that can be used to remove it (hook style) otherwise it will live forever.


The argument about having invalid Library object after "remove" is good though.
I guess a good way forward here is:
- Make the lifetime linked to the Library object (so no need for this function)
- We can remove the global dict `_impls_for_existing_libraries` and don't populate it in the init.
- remove the `extend_library` function altogether and simply make the `kind` argument optional for `Library` with a default to `IMPL`.